HW0/P4/P4.py

```diff
@@ -40,37 +40,70 @@
     # Read the observation array and plot it (Part 2)
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    measurement = np.loadtxt('P4_measurements.txt', delimiter = ',', unpack = True)
+    ax.plot(measurement[0] * (1/rx), measurement[1] * (1/ry), measurement[2] * (1/rz), '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
-
     # Initial conditions for s0
-    # Compute the rest of sk using Eq (1)
+    s0 = np.matrix([0.,0.,2.,15.,3.5,4]).T
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    # Compute the rest of sk using Eq (1)
+    a = np.matrix([0,0,0,0,0,g*dt]).T
+    A = np.matrix(np.zeros([6,6]))
+    A[0,0] = A[1,1] = A[2,2] = 1
+    A[0,3] = A[1,4] = A[2,5] = dt
+    A[3,3] = A[4,4] = A[5,5] = 1 - c*dt
+    
+    pred = np.matrix(np.zeros([6, K]))
+
+    pred[:, 0] = s0
+    for i in np.arange(K):
+	if i == 0: pred[:, 0] = s0
+	else: pred[:, i] = (A * pred[:, i-1]) + a
+ 
+    pred = np.array(pred)
+  
+    ax.plot(pred[0], pred[1], pred[2], '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.matrix(np.identity(6))
+    B *= np.matrix([bx, by, bz, bvx, bvy, bvz]).T
+
+    C = np.matrix(np.zeros([3,6]))
+    C[0,0], C[1,1], C[2,2] = rx, ry, rz
 
     # Initial conditions for s0 and Sigma0
-    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    Sigma0 = 0.01*np.identity(6)
+    
+    Sigmak, sk, m = Sigma0, s0, np.matrix(measurement)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    predictS = lambda A, sk, a: A*sk + a
```

Comment:
I found your code to be quite clean! Nice work. I especially liked the use of lambda functions to compute the rest of sk using equations 2, 3, 4 and 5.